### ISSUE TITLE
Design/execution: train qwen-lean v3 with incremental SFT and anti-collapse gates

### ISSUE THREAD

**Issue by murillo128** [murillo128/qwen-lean#83]
## Goal

Train the next general qwen-lean executor from `Qwen/Qwen3.5-4B-Base` using **Dataset v3 from #82**, with an objective and execution plan designed to improve Lean capability **without collapsing Base search diversity**.

This issue is blocked on #82. Do not start the v3 training run from Dataset v2 or from an incomplete Dataset-v3 package.

The central hypothesis is that the v2 failure was not simply “too much training”: Q1 already collapsed to 0/415 on the opened fresh-composition diagnostic, while the Q4-specific inference canary proved the adapter was faithfully active. The v3 training strategy therefore changes both the training examples and the anti-collapse gates.

## Training interface

Dataset v3 should expose two compatible task families:

1. **whole proof**

```text
theorem -> full proof
```

2. **incremental proof continuation**

```text
theorem + verified proof prefix -> remaining continuation
```

Train one executor on both. The model must remain able to start a proof from scratch and to continue a promising trajectory supplied by a future planner/controller.

Do not convert this issue into a tactic-state-only prover. Explicit Lean proof-state conditioning may be explored later; v3 initially uses source theorem/context + proof prefix as defined by #82.

## Stage 0 — freeze evaluation and anti-collapse diagnostics before optimizer updates

Before training:

- bind the exact accepted Dataset-v3 manifest/hashes from #82;
- freeze the new v3 validation and sealed v3 test memberships;
- evaluate the unchanged Base on the v3 validation workloads and record candidate-level outputs;
- define a small **anti-collapse validation canary** drawn only from v3 validation-safe material, not the sealed v3 test;
- retain historical miniF2F and v2 fresh-composition results as diagnostic comparators only; the already-opened v2 final test must not drive v3 checkpoint selection.

The anti-collapse canary must measure at least:

- Lean-valid candidate rate / solved-within-k;
- EOS vs token-limit rate;
- generated-token distribution;
- first proof construct / major proof-form distribution;
- normalized output/template diversity;
- task-level output overlap with Base;
- direct logit/forward drift where practical.

## Stage 1 — bounded early training, not an immediate full pass

Do **not** launch another 22k-step full pass as the first experiment.

Run a bounded representative v3 training stream with dense early checkpoints, for example around:

```text
Base
100
250
500
1k
2k
4k optimizer steps
```

The exact boundaries may be adjusted slightly for implementation convenience, but the purpose is to locate the onset of useful Lean learning versus distribution narrowing well before the old Q1 boundary (~5.7k steps).

At every retained checkpoint run the same frozen anti-collapse canary. Publish a trajectory table including:

- Lean success/coverage;
- valid-candidate density;
- output diversity;
- `exact`/constructor and other proof-form frequencies;
- EOS rate and length distribution;
- drift from Base;
- train loss and optimizer health.

A checkpoint that improves imitation loss while catastrophically narrowing proof behavior is a **failed checkpoint**, not a candidate to continue training from merely because validation CE improved.

## Stage 2 — training objective / Base preservation

The v2 objective was plain target-only whole-proof SFT with LR `1e-4`, no weight decay, and a highly expressive LoRA across attention, Gated DeltaNet, and MLP. V3 must explicitly control behavioral drift.

Requirements:

- use the whole-proof + incremental Dataset-v3 examples and theorem-normalized example weights from #82;
- use a materially less aggressive optimization setup than v2 unless bounded evidence supports otherwise;
- include an explicit **Base-preservation mechanism** or equivalent behavioral constraint rather than relying only on the fact that the base weights are frozen;
- preserve the ability to monitor the preservation term separately from Lean SFT loss;
- keep the approach feasible on the project Ada-class GPU path.

Preferred design: a KL/distillation-style preservation term against the frozen Base on a fixed anchor set, using cached/reference logits or another bounded implementation that does not require two full 4B models resident simultaneously. If a different mechanism is chosen, document why it is an equivalent guard against v2-style catastrophic distribution sharpening.

Do not hard-code a large hyperparameter sweep. Use the smallest bounded comparison needed to choose a preservation strength / learning rate that shows both:

1. measurable Lean learning;
2. no catastrophic anti-collapse regression.

## Stage 3 — continue only after an anti-collapse PASS

Only if a bounded checkpoint demonstrates useful Lean learning without reproducing the v2 narrowing may the run continue toward the intended Dataset-v3 training budget.

Continuation rules:

- preserve the chosen objective/config unchanged after the gate;
- retain denser checkpoints around any observed transition region, then reduce frequency once behavior is stable;
- use validation only for stopping/checkpoint selection;
- never inspect the sealed v3 test to choose training depth, LR, preservation strength, or whole/incremental weighting;
- no test-driven checkpoint reselection.

The final SFT checkpoint should be chosen by a **multi-signal validation rule** that prioritizes actual Lean generation capability and anti-collapse behavior, not training loss alone. At minimum it must reject any checkpoint with a material broad-regression/collapse gate failure even if one validation aggregate improves.

## Stage 4 — final evaluation

After checkpoint freeze, evaluate Base and frozen qwen-lean-v3 on the sealed v3 test under the same matched sampling/verifier budget.

Report:

- pass@1/pass@4/pass@8 and solved-within-8;
- paired task-level gains/losses;
- Lean-valid candidate density;
- EOS/length distributions;
- normalized proof/template diversity;
- proof-form histogram;
- overlap versus Base;
- capability-gap breakdown analogous to useful parts of #78 Stage 9.

Also report historical miniF2F/v2 diagnostic results separately, clearly marked as already-seen/non-selection benchmarks where applicable.

## Success criterion

V3 is not required to beat Base on every metric. It **is** required to demonstrate that Lean specialization no longer causes the catastrophic v2 behavior where a trained checkpoint collapses an independently held-out composition workload to zero while emitting near-uniform short templates.

A successful v3 SFT should show:

- meaningful verified Lean learning relative to Base on v3 validation/test;
- no catastrophic loss of Base-held theorem coverage;
- materially healthier output/proof-form diversity than v2 Q1/Q4;
- useful incremental-continuation capability suitable for later planner guidance.

If no bounded configuration can achieve Lean learning without immediate collapse, stop rather than scaling the run. That would be evidence that supervised imitation remains the wrong primary executor objective and should trigger a verifier-driven training design rather than another larger SFT.

## Verifier-driven post-training boundary

Do **not** fold a large GRPO/RLVR campaign into the first v3 SFT experiment. First establish a non-collapsed incremental executor and measure its downstream reward density.

If v3 SFT passes, the next step may use Lean verification for verifier-filtered self-training / preference optimization / RLVR. That follow-up should reuse the executor and can be added as a later amendment or follow-up issue once reward density is measured. The immediate purpose here is to fix the corpus/objective failure exposed by v2 in a scientifically isolatable way.

## Preserve useful v2 engineering

Reuse the validated infrastructure from #78 / PR #79 where it is model-agnostic and correct, especially:

- exact model/revision binding;
- Qwen3.5 HF/PEFT ↔ vLLM LoRA namespace/parity gate;
- Lean verification and retained candidate evidence;
- checkpoint/adaptor hash binding;
- candidate-level pass@k/paired analysis infrastructure.

Do not inherit v2's corpus serialization or training objective merely because the surrounding evaluator code is reusable.

## Explicit non-goals

- no Mathia training;
- no planner training;
- no Riemann-specific specialization;
- no use of opened v2 final-test failures as v3 training examples;
- no broad model-family tournament inside this issue;
- no automatic promotion of a checkpoint that fails anti-collapse gates.

## Dependency / completion

**Blocked by #82.**

This issue completes when a frozen qwen-lean-v3 SFT executor has been trained and independently evaluated on the sealed Dataset-v3 test, with evidence showing whether incremental/source-preserving SFT plus Base preservation solves the v2 mode-collapse failure. If it does, that checkpoint becomes the executor candidate for later planner/verifier-driven work; if it does not, stop and carry the negative result into a verifier-driven training redesign rather than silently retrying the same SFT pattern.

**Comment by murillo128** [comment]
## Complete bounded trajectory published — SFT stopped

Exact target: `aabd607878b29d9d61f1d73dd9338a7d3250003a` (draft PR #85).

All retained evaluations are now complete: Base plus C0/C1/C2/C3 at optimizer steps 100/250/500. Each checkpoint used the frozen 96-task / 768-candidate canary with native context, no truncation, no silent exclusions, local vLLM generation, and local Lean verification.

Solved / verified trajectory:
- C0: step 100 `6/6`, step 250 `4/5`, step 500 `6/8`; step-500 mean anchor KL `4.9639207203`;
- C1: `2/2`, `1/1`, `0/0`; step-500 KL `0.4560910587`;
- C2: `2/2`, `0/0`, `0/0`; step-500 KL `0.4870596997`;
- C3 rescue: `0/0`, `2/2`, `0/0`; step-500 KL `0.3779636238`.

Every one of the 12 retained checkpoints failed the hard repeated-template collapse gate in the whole-proof lane and is ineligible. C1, C2, and C3 each failed the positive step-500 gate. The Base-preservation term clearly constrained direct anchor KL, but did not preserve healthy search behavior.

Frozen outcome: SFT `stopped`; no selected/frozen checkpoint; no optimizer updates beyond 500; no 1k/2k/4k/8k continuation; sealed Dataset-v3 test untouched. The negative result carries into a separate verifier-driven redesign. The newly cited replay precedent is useful follow-up design input but does not reopen this execution.

Evidence SHA-256:
- aggregate trajectory: `228e21d0821b36b6d285427bdc8cf40c2f5b905f4820f0649e70d6211190def3`;
- validation plot: `f97f8900d85817d79b132154082d6f4677bac18a3a7a815cb76f9b4a1d39488e`;
- training plot: `d5af8d05e8886049cfdddce26654f4d08394e13246fe6d89f561bd7fcde58bf9`.

Validation: 399 passed, 2 known optional-runtime skips; Python compilation, diff, JSON, and secret-pattern checks passed.

Next and final execution gate: independent review of exact SHA `aabd607878b29d9d61f1d73dd9338a7d3250003a`.

**Comment by murillo128** [comment]
## Independent-review corrections published

The first final review of `aabd607878b29d9d61f1d73dd9338a7d3250003a` returned `FAIL` for two material reporting/gate defects. Both are corrected at exact target `1386f1ad37994565adf476445e366ac571b1a66c` in draft PR #85.

Corrections:
- repeated-template collapse now checks every normalized template across theorem IDs, not only the most frequent template; a regression test covers a wide but less frequent violating template;
- every compact canary was recomputed from retained raw candidates and now records per-template theorem/occurrence/verification counts plus p90;
- the canonical trajectory contains 637 denominator-bearing validation rows and all 2,000 raw per-step objective rows;
- the deterministic 11-view plot package now covers separate whole/incremental solved@8 and density, Base retention, EOS, median/p75/p90 length, diversity, nine major first-construct categories, direct/branching/deep capability, anchor KL drift, and raw SFT/preservation objectives.

The corrected gate found one or more whole-proof violating templates at every retained checkpoint. Some checkpoints had multiple violations, but all prior eligibility and scientific stop decisions remain unchanged. The complete bounded trajectory is still:
- C0 solved/verified: `6/6`, `4/5`, `6/8`; step-500 KL `4.9639207203`;
- C1: `2/2`, `1/1`, `0/0`; KL `0.4560910587`;
- C2: `2/2`, `0/0`, `0/0`; KL `0.4870596997`;
- C3: `0/0`, `2/2`, `0/0`; KL `0.3779636238`.

Frozen outcome remains SFT `stopped`: no selected/frozen checkpoint, no updates beyond 500, and the sealed Dataset-v3 test remains untouched.

Corrected aggregate trajectory SHA-256: `a95fba2983511d5b6206893e597ab7641276717c4786d215fbf0e32405257727`; it binds every plot hash.

Validation: 400 passed, 2 known optional-runtime skips. Compilation, diff, JSON, self-hash, byte-for-byte regeneration, and secret scans passed.

Next gate: fresh independent final review of exact SHA `1386f1ad37994565adf476445e366ac571b1a66c`.

**Comment by murillo128** [comment]
## Independent final review — PASS

Exact reviewed target: `1386f1ad37994565adf476445e366ac571b1a66c`.

Verdict: `PASS`. This serves as final technical review; no material findings remain. It is not merge authorization.

The fresh read-only reviewer independently:
- recomputed the every-template collapse gate and matched every violating-template list and eligibility decision;
- re-compacted all 13 raw canaries byte-for-byte to committed evidence;
- verified Base plus 12 checkpoint completeness, all adapter/training-stream/drift/evidence bindings, and exact 500-update boundaries;
- matched 637 denominator-bearing validation rows and 2,000 finite raw objective rows;
- regenerated all 11 plots and the aggregate byte-for-byte from committed evidence;
- confirmed no truncation, silent exclusions, later checkpoints, or sealed-test access;
- ran the full suite: 400 passed, 2 declared optional-runtime skips.

The bounded-SFT negative result is technically supported and safely progresses to user review. No merge or auto-merge action was taken.
